### ISSUE TITLE
FieldArray fields.getAll does not reflect redux state 

### ISSUE THREAD

**Issue by 19majkel94** [erikras/redux-form#3679]
This bug is related to #2355.

I want to create a dynamic list of inputs where there is always empty input on the bottom:
```ts
<CustomField
  key={index}
  name={name}
  component={this.renderDynamicIngredientInput}
  onUpdate={(event: React.ChangeEvent<HTMLInputElement>) => {
    //console.log(fields.getAll());
    const lastField = fields.getAll()[fields.length - 1];
    //console.log(event.target.value.length, lastField.length, lastField, fields.length);
    if (event.target.value.length > 0 && lastField.length > 0) {
      fields.push("");
    }
  }}
/>
```
```ts
<input
  {...input}
  {...props}
  onChange={event => {
    input.onChange(event);
    onUpdate(event);
  }}
/>
```
The view:
![chrome_2017-12-08_22-34-39](https://user-images.githubusercontent.com/10618781/33786392-48bb1f3a-dc69-11e7-81af-eade03064f8d.png)

The store state:
![chrome_2017-12-08_22-34-33](https://user-images.githubusercontent.com/10618781/33786410-5b65973c-dc69-11e7-9de8-7080da33a9ca.png)

The console.log with `getAll`:
![chrome_2017-12-08_22-34-48](https://user-images.githubusercontent.com/10618781/33786423-660cc746-dc69-11e7-9642-847e65fe3c51.png)

I'm using `"redux-form": "^7.2.0",` 😉 

Basically it's get updated only on first type after focus in input.

**Comment by 19majkel94** [comment]
I don't know why it haven't want to work but it didn't as you can see on screenshots.
Now I'm using `fields.length` and it works 😃 

```ts
onType={(event: React.ChangeEvent<HTMLInputElement>) => {
  if (
    index === fields.length - 1 &&
    event.target.value.length > 0
  ) {
    fields.push("");
  } else if (event.target.value.length === 0) {
    fields.remove(index);
  }
}}
```
So I'm closing it for now. If it happens again, I will create a sandbox for it 😉

**Comment by giocodes** [comment]
I ran through this problem and `fields.getAll()` was always one step behind the store values, here is a solution that worked for us:
```javascript
onCopyFieldValues(fields, options) {
    let documents = [];
    fields.forEach((item, idx) => documents.push(fields.get(idx)));
    const newFields = copyFieldValues(documents, options); // running custom changes to the values
    this.props.change('documents', newFields);
  }
```

**TLTR:** fields.getAll() didn't work but calling `fields.get(<index>)` with `fields.forEach()` did the trick.

**Comment by jarekb84** [comment]
@giocodes Thanks for that solution. Solved my issue of passing in fields.getAll() as a prop to a pure component and not seeing re-renders when data updated. Anyone know why .getAll() behaves that way? Is it just a timing thing (getting data before store is updated?)

**Comment by JesseTheNoob** [comment]
> I ran through this problem and `fields.getAll()` was always one step behind the store values, here is a solution that worked for us:
> 
> ```js
> onCopyFieldValues(fields, options) {
>     let documents = [];
>     fields.forEach((item, idx) => documents.push(fields.get(idx)));
>     const newFields = copyFieldValues(documents, options); // running custom changes to the values
>     this.props.change('documents', newFields);
>   }
> ```
> 
> **TLTR:** fields.getAll() didn't work but calling `fields.get(<index>)` with `fields.forEach()` did the trick.

@giocodes, I tried using `fields.forEach((item, idx) => documents.push(fields.get(idx)));` but it is still 1 behind when I console logged the documents. pls help

**Comment by jKh98** [comment]
> > I ran through this problem and `fields.getAll()` was always one step behind the store values, here is a solution that worked for us:
> > ```js
> > onCopyFieldValues(fields, options) {
> >     let documents = [];
> >     fields.forEach((item, idx) => documents.push(fields.get(idx)));
> >     const newFields = copyFieldValues(documents, options); // running custom changes to the values
> >     this.props.change('documents', newFields);
> >   }
> > ```
> > 
> > 
> > **TLTR:** fields.getAll() didn't work but calling `fields.get(<index>)` with `fields.forEach()` did the trick.
> 
> @giocodes, I tried using `fields.forEach((item, idx) => documents.push(fields.get(idx)));` but it is still 1 behind when I console logged the documents. pls help

It have been stuck on this for a couple of days now. Were you able to solve it?

Thanks in advance

**Comment by giocodes** [comment]
@JesseTheNoob @jKh98  sorry I haven't used this in a loooong time, if you setup a codesandbox to reproduce the issue I'll be happy to take a look.
